File: tools/test-runtime-stores.py
"""Host tests execute production stores; Android stubs model memory-before-disk failure.

No device, root executable, installation, provider call or Global setting is used.
Provider tests compile its actual entry prefix and allowedCaller method, excluding
unrelated routes. This verifies policy ordering, not Android Binder integration.
"""
from pathlib import Path
import subprocess
import sys

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "outputs/native-fixes/runtime-store-tests"
SRC = ROOT / "android/app/src/main/java/ls/augment/com"
STUBS = {
    "android/content/SharedPreferences.java": """package android.content;
import java.util.Map;
public interface SharedPreferences {
 Map<String,?> getAll(); String getString(String k,String d); boolean getBoolean(String k,boolean d);
 int getInt(String k,int d); long getLong(String k,long d); Editor edit();
 interface Editor { Editor putString(String k,String v); Editor putBoolean(String k,boolean v);
 Editor putInt(String k,int v); Editor putLong(String k,long v); Editor putFloat(String k,float v);
 Editor clear(); boolean commit(); }
}""",
    "android/content/Context.java": """package android.content;
public abstract class Context {
 private final android.content.pm.PackageManager packages = new android.content.pm.PackageManager();
 public abstract SharedPreferences getSharedPreferences(String name,int mode);
 public abstract ContentResolver getContentResolver();
 public android.content.pm.PackageManager getPackageManager(){return packages;}
}""",
    "android/content/ContentResolver.java": """package android.content;
public class ContentResolver { public int notifications;
 public void notifyChange(android.net.Uri uri,Object observer){notifications++;} }
""",
    "android/content/pm/PackageManager.java": """package android.content.pm;
public class PackageManager { public final java.util.Map<Integer,String[]> packages=new java.util.HashMap<>();
 public String[] getPackagesForUid(int uid){return packages.get(uid);} }
""",
    "android/net/Uri.java": """package android.net;
public class Uri { public static Uri parse(String value){return new Uri();} }
""",
    "android/os/Bundle.java": """package android.os;
public class Bundle { private final java.util.Map<String,Object> data=new java.util.HashMap<>();
 public void putBoolean(String k,boolean v){data.put(k,v);} public boolean getBoolean(String k){return getBoolean(k,false);}
 public boolean getBoolean(String k,boolean d){Object v=data.get(k);return v instanceof Boolean?(Boolean)v:d;}
 public void putString(String k,String v){data.put(k,v);} public String getString(String k){return getString(k,null);}
 public String getString(String k,String d){Object v=data.get(k);return v instanceof String?(String)v:d;}
 public void putInt(String k,int v){data.put(k,v);} public int getInt(String k){return getInt(k,0);}
 public int getInt(String k,int d){Object v=data.get(k);return v instanceof Integer?(Integer)v:d;}
 public void putLong(String k,long v){data.put(k,v);} public long getLong(String k){Object v=data.get(k);return v instanceof Long?(Long)v:0;}
} """,
    "android/os/Binder.java": """package android.os;
public class Binder { public static int uid=12000,clears,restores;
 public static int getCallingUid(){return uid;} public static long clearCallingIdentity(){clears++;int old=uid;uid=Process.myUid();return old;}
 public static void restoreCallingIdentity(long old){restores++;uid=(int)old;} }
""",
    "android/os/Process.java": """package android.os;
public class Process { public static final int SYSTEM_UID=1000,ROOT_UID=0;public static int myUid(){return 10055;} }
""",
    "ls/augment/com/RootShell.java": """package ls.augment.com;
final class RootShell { static Result next=new Result(0,"null\\nnull\\nnull",false);static int calls;
 static Result run(String c,String i,long t,int m){calls++;return next;}
 static final class Result { final int exitCode;final String output;final boolean timedOut;
 Result(int code,String out,boolean timeout){exitCode=code;output=out;timedOut=timeout;}
 boolean isSuccess(){return exitCode==0&&!timedOut;} }
}""",
    "ls/augment/com/FrameworkConfigSync.java": """package ls.augment.com;
final class FrameworkConfigSync { static int requests;static void request(){requests++;} }
""",
    "ls/augment/com/BootJobService.java": """package ls.augment.com;
final class BootJobService { static int requests;static boolean accepted=true;
 static boolean scheduleHiddenRefresh(android.content.Context context){
 if(android.os.Binder.getCallingUid()!=android.os.Process.myUid())throw new AssertionError("recovery scheduled before identity clear");
 requests++;return accepted;} }
""",
    "ls/augment/com/BuildConfig.java": """package ls.augment.com;
final class BuildConfig { static boolean DEBUG=true; }
""",
    "ls/augment/com/HookTargetRegistry.java": """package ls.augment.com;
final class HookTargetRegistry { static boolean contains(String name){return "com.android.systemui".equals(name);} }
""",
}

provider = (SRC / "LSConfigProvider.java").read_text(encoding="utf-8")
call_start = provider.index("        Bundle result = new Bundle();", provider.index("public Bundle call("))
call_end = provider.index("        if (ShoulderQuickSwitchPolicy.CALL", call_start)
prefix = provider[call_start:call_end]
assert prefix.index("!allowedCaller()") < prefix.index('"runtime_snapshot"')
crash_prefix = prefix.split('if ("crash_fuse".equals(method))', 1)[1]
assert crash_prefix.index("Binder.getCallingUid() != Process.SYSTEM_UID") < crash_prefix.index("Binder.clearCallingIdentity()")
recovery_prefix = prefix.split('if ("recovery_refresh".equals(method))', 1)[1].split('if (!allowedCaller())', 1)[0]
assert recovery_prefix.index("Binder.getCallingUid() != Process.ROOT_UID") < recovery_prefix.index("Binder.clearCallingIdentity()")
gate_start = provider.index("    private boolean allowedCaller()")
gate_end = provider.index("    private Bundle recordRapidRoute", gate_start)
gate = provider[gate_start:gate_end]
STUBS["ls/augment/com/ProviderEntry.java"] = """package ls.augment.com;
import android.content.Context;import android.os.Bundle;import android.os.Binder;import android.os.Process;
final class ProviderEntry { private final Context context;ProviderEntry(Context c){context=c;}Context getContext(){return context;}
public Bundle call(String method,String arg,Bundle extras){
""" + prefix + "return result;}\n" + gate + "}\n"

files = []
for relative, content in STUBS.items():
    path = OUT / "src" / relative
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")
    files.append(path)
classes = OUT / "classes"
classes.mkdir(parents=True, exist_ok=True)
files += [SRC / name for name in ("DurablePreferences.java", "RuntimeStateStore.java", "CrashFuseStore.java", "RemoteConfig.java")]
files += [ROOT / "tools" / name for name in ("TestRuntimeStores.java", "TestProviderEntry.java")]
subprocess.run(["javac", "-encoding", "UTF-8", "--release", "17", "-d", str(classes), *map(str, files)], check=True)
cases = ("runtime_success", "runtime_failed_retry", "runtime_failed_snapshot", "runtime_idempotent",
         "fuse_unmigrated", "fuse_failed_reset", "fuse_failed_migration", "fuse_bad_migration",
         "fuse_third_crash", "fuse_failed_arm", "fuse_stable", "fuse_repeated_session",
         "fuse_failed_counter", "durable_failed_key_leak", "runtime_cold_reload", "fuse_cold_reload")
if '--provider-only' in sys.argv:
    cases = ()
failed = []
for case in cases:
    result = subprocess.run(["java", "-cp", str(classes), "ls.augment.com.TestRuntimeStores", case], check=False)
    if result.returncode:
        failed.append(case)
result = subprocess.run(["java", "-cp", str(classes), "ls.augment.com.TestProviderEntry"], check=False)
if result.returncode:
    failed.append("provider_entry")
if cases:
    print(f"Runtime store scenarios: {len(cases) - len([x for x in failed if x != 'provider_entry'])}/{len(cases)} passed")
if failed:
    raise SystemExit("Failed: " + ", ".join(failed))

File: tools/TestRuntimeStores.java
package ls.augment.com;

import android.content.Context;
import android.content.ContentResolver;
import android.content.SharedPreferences;
import android.os.Bundle;
import java.util.HashMap;
import java.util.Map;

/** Runs against production stores with a host-side SharedPreferences fault model. */
public final class TestRuntimeStores {
    private static final String A = "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa";
    private static final String B = "bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb";
    public static void main(String[] args) {
        FakeContext context = new FakeContext();
        FakePreferences runtime = context.prefs("ls_augment_runtime_v2");
        FakePreferences fuse = context.prefs("ls_augment_crash_fuse_v2");
        String test = args[0];
        switch (test) {
            case "runtime_success":
                check(RuntimeStateStore.publishHidden(context, "999:com.test.app", "ALL_HIDDEN").isSuccess(), "publish succeeds");
                check("999:com.test.app".equals(runtime.disk.get(RemoteConfig.HIDDEN)), "hidden targets durable");
                check(context.resolver.notifications == 1 && FrameworkConfigSync.requests == 1, "notify after durable publication");
                check(RuntimeStateStore.snapshot(context).getBoolean("ok"), "snapshot available");
                break;
            case "runtime_failed_retry":
                runtime.failNext = 1;
                check(!RuntimeStateStore.publishHidden(context, "0:com.test.app", "ALL_HIDDEN").isSuccess(), "first write fails");
                check(context.resolver.notifications == 0 && FrameworkConfigSync.requests == 0, "failed state not published");
                check(RuntimeStateStore.publishHidden(context, "0:com.test.app", "ALL_HIDDEN").isSuccess(), "same-value retry succeeds");
                check(runtime.commits >= 2 && "0:com.test.app".equals(runtime.disk.get(RemoteConfig.HIDDEN)), "same-value retry really reaches disk");
                break;
            case "runtime_failed_snapshot":
                check(RuntimeStateStore.publishHidden(context, "0:com.old.app", "ALL_HIDDEN").isSuccess(), "baseline durable");
                runtime.failNext = 1;
                check(!RuntimeStateStore.publishHidden(context, "999:com.new.app", "MIXED").isSuccess(), "second write fails");
                check("0:com.old.app".equals(RuntimeStateStore.snapshot(context).getString(RemoteConfig.HIDDEN)), "readers retain last durable hidden set");
                check("ALL_HIDDEN".equals(RuntimeStateStore.snapshot(context).getString(RemoteConfig.TILE)), "readers retain last durable tile");
                break;
            case "runtime_idempotent":
                RuntimeStateStore.publishHidden(context, "0:com.test.app", "ALL_HIDDEN");
                int commits = runtime.commits;
                RuntimeStateStore.publishHidden(context, "0:com.test.app", "ALL_HIDDEN");
                check(runtime.commits == commits, "confirmed unchanged snapshot avoids disk write");
                check(!RuntimeStateStore.publishHidden(context, null, "ALL_HIDDEN").isSuccess(), "invalid hidden value rejected");
                check(!RuntimeStateStore.publishHidden(context, "", "lower case").isSuccess(), "invalid tile rejected");
                break;
            case "fuse_unmigrated":
                check(CrashFuseStore.isFused(context), "unmigrated is fail-closed");
                check(!CrashFuseStore.refresh(context, request(A, "arm")).getBoolean("ok"), "unmigrated cannot arm");
                RootShell.next = new RootShell.Result(127, "denied", false);
                check(!CrashFuseStore.importLegacy(context) && CrashFuseStore.isFused(context), "Root unavailable cannot bypass migration");
                break;
            case "fuse_failed_reset":
                RootShell.next = new RootShell.Result(0, "0\n3\n1", false);
                check(CrashFuseStore.importLegacy(context) && CrashFuseStore.isFused(context), "tripped legacy imported");
                fuse.failNext = 1;
                check(!CrashFuseStore.reset(context), "reset failure reported");
                check(CrashFuseStore.isFused(context), "failed reset cannot untrip in memory");
                check("FUSED".equals(CrashFuseStore.refresh(context, request(A, "arm")).getString("state")), "later refresh cannot commit failed reset as success");
                break;
            case "fuse_failed_migration":
                RootShell.next = new RootShell.Result(0, "null\nnull\nnull", false);
                fuse.failNext = 1;
                check(!CrashFuseStore.importLegacy(context), "migration disk failure reported");
                check(CrashFuseStore.isFused(context), "failed migration remains blocked");
                check(CrashFuseStore.importLegacy(context), "migration retry succeeds");
                check(Boolean.TRUE.equals(fuse.disk.get("migrated")) && fuse.commits >= 2, "migration retry reaches disk");
                break;
            case "fuse_bad_migration":
                RootShell.next = new RootShell.Result(0, "0\nnot-a-number\n0", false);
                check(!CrashFuseStore.importLegacy(context) && CrashFuseStore.isFused(context), "malformed legacy blocked");
                RootShell.next = new RootShell.Result(0, "0\n-1\n0", false);
                check(!CrashFuseStore.importLegacy(context) && CrashFuseStore.isFused(context), "negative legacy blocked");
                break;
            case "fuse_third_crash":
                RootShell.next = new RootShell.Result(0, "1\n2\n0", false);
                check(CrashFuseStore.importLegacy(context), "pending legacy imported");
                Bundle result = CrashFuseStore.refresh(context, request(A, "arm"));
                check(result.getInt("attempts") == 3 && "FUSED".equals(result.getString("state")), "third interrupted session trips");
                result = CrashFuseStore.refresh(context, request(A, "clearAttempts", "clearPending", "arm"));
                check("FUSED".equals(result.getString("state")), "clearing counters cannot clear tripped fuse");
                break;
            case "fuse_failed_arm":
                check(CrashFuseStore.reset(context), "reset durable");
                fuse.failNext = 1;
                check(!CrashFuseStore.refresh(context, request(A, "arm")).getBoolean("ok"), "failed arm returns unknown");
                check(!Boolean.TRUE.equals(fuse.disk.get("pending")), "failed arm is not durable");
                Bundle armed = CrashFuseStore.refresh(context, request(A, "arm"));
                check(armed.getBoolean("ok") && "ARMED".equals(armed.getString("state"))
                        && Boolean.TRUE.equals(fuse.disk.get("pending")), "retry arms only after durable marker");
                break;
            case "fuse_stable":
                check(CrashFuseStore.reset(context), "reset durable");
                check("ARMED".equals(CrashFuseStore.refresh(context, request(A, "arm")).getString("state")), "first session arms");
                CrashFuseStore.refresh(context, request(A, "clearPending", "clearAttempts"));
                Bundle next = CrashFuseStore.refresh(context, request(B));
                check("READY".equals(next.getString("state")) && next.getInt("attempts") == 0, "stable session not counted as crash");
                break;
            case "fuse_repeated_session":
                CrashFuseStore.reset(context);
                CrashFuseStore.refresh(context, request(A, "arm"));
                Bundle once = CrashFuseStore.refresh(context, request(B, "arm"));
                check(once.getInt("attempts") == 1, "new session counts pending interruption once");
                for (int i = 0; i < 5; i++)
                    check(CrashFuseStore.refresh(context, request(B, "arm")).getInt("attempts") == 1,
                            "same session does not double count");
                break;
            case "fuse_failed_counter":
                CrashFuseStore.reset(context);
                CrashFuseStore.refresh(context, request(A, "arm"));
                fuse.failNext = 1;
                check(!CrashFuseStore.refresh(context, request(B, "arm")).getBoolean("ok"), "failed count is not acknowledged");
                Bundle counted = CrashFuseStore.refresh(context, request(B, "arm"));
                check(counted.getInt("attempts") == 1 && (Integer) fuse.disk.get("attempts") == 1,
                        "retry counts original interrupted session exactly once");
                break;
            case "durable_failed_key_leak":
                DurablePreferences store = new DurablePreferences(runtime);
                check(store.edit().putString("original", "yes").commit(), "baseline");
                runtime.failNext = 1;
                check(!store.edit().putString("failed", "no").commit(), "optimistic failed key inserted");
                check(!store.contains("failed") && runtime.memory.containsKey("failed"), "confirmed ignores optimistic memory");
                check(store.edit().putString("next", "yes").commit(), "different later write succeeds");
                check(!runtime.disk.containsKey("failed") && "yes".equals(runtime.disk.get("original")), "full replacement prevents failed key leakage");
                break;
            case "runtime_cold_reload":
                RuntimeStateStore.publishHidden(context, "0:com.old.app", "ALL_HIDDEN");
                runtime.failNext = 1;
                RuntimeStateStore.publishHidden(context, "999:com.new.app", "MIXED");
                coldReload(RuntimeStateStore.class, runtime);
                check("0:com.old.app".equals(RuntimeStateStore.snapshot(context).getString(RemoteConfig.HIDDEN)),
                        "cold process reloads durable state after failed write");
                break;
            case "fuse_cold_reload":
                CrashFuseStore.reset(context);
                CrashFuseStore.refresh(context, request(A, "arm"));
                coldReload(CrashFuseStore.class, fuse);
                Bundle reloaded = CrashFuseStore.refresh(context, request(B, "arm"));
                check(reloaded.getInt("attempts") == 1 && "ARMED".equals(reloaded.getString("state")),
                        "cold process counts persisted pending marker once");
                break;
            default: throw new AssertionError(test);
        }
        System.out.println(test + ": OK");
    }

    private static Bundle request(String session, String... flags) {
        Bundle value = new Bundle(); value.putString("session", session);
        for (String flag : flags) value.putBoolean(flag, true);
        return value;
    }
    private static void check(boolean value, String message) { if (!value) throw new AssertionError(message); }

    private static void coldReload(Class<?> store, FakePreferences preferences) {
        preferences.memory.clear(); preferences.memory.putAll(preferences.disk);
        try {
            java.lang.reflect.Field field = store.getDeclaredField("durable");
            field.setAccessible(true); field.set(null, null);
        } catch (ReflectiveOperationException error) { throw new AssertionError(error); }
    }

    static final class FakeContext extends Context {
        final Map<String, FakePreferences> stores = new HashMap<>();
        final ContentResolver resolver = new ContentResolver();
        FakePreferences prefs(String name) { return stores.computeIfAbsent(name, unused -> new FakePreferences()); }
        @Override public SharedPreferences getSharedPreferences(String name, int mode) {
            check(mode == 0, "all preferences must remain app-private"); return prefs(name);
        }
        @Override public ContentResolver getContentResolver() { return resolver; }
    }

    static final class FakePreferences implements SharedPreferences {
        final Map<String, Object> memory = new HashMap<>(), disk = new HashMap<>();
        int failNext, commits;
        public Map<String, ?> getAll() { return new HashMap<>(memory); }
        public String getString(String key, String fallback) { return (String) memory.getOrDefault(key, fallback); }
        public boolean getBoolean(String key, boolean fallback) { return (Boolean) memory.getOrDefault(key, fallback); }
        public int getInt(String key, int fallback) { return (Integer) memory.getOrDefault(key, fallback); }
        public long getLong(String key, long fallback) { return (Long) memory.getOrDefault(key, fallback); }
        public Editor edit() {
            return new Editor() {
                final Map<String, Object> pending = new HashMap<>();
                boolean clear;
                public Editor putString(String key, String value) { pending.put(key, value); return this; }
                public Editor putBoolean(String key, boolean value) { pending.put(key, value); return this; }
                public Editor putInt(String key, int value) { pending.put(key, value); return this; }
                public Editor putLong(String key, long value) { pending.put(key, value); return this; }
                public Editor putFloat(String key, float value) { pending.put(key, value); return this; }
                public Editor clear() { clear = true; return this; }
                public boolean commit() {
                    commits++;
                    if (clear) memory.clear();
                    memory.putAll(pending); // AOSP commitToMemory happens before disk completion.
                    if (failNext > 0) { failNext--; return false; }
                    disk.clear(); disk.putAll(memory); return true;
                }
            };
        }
    }
}
